[Zack-Tay] iP

Zack-Tay opened the pull request:
**Fredricksen**
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Fredricksen frees your mind of having to remember things you need to do. It's,

- text-based
* easy to learn
+ ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2324s2.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

Zack-Tay committed:
Add Fredricksen.java and completed level-0

Zack-Tay committed:
Add echo and bye commands to Fredricksen.java and completed Level-1

Zack-Tay committed:
add add task to list feature to Fredricksen.java

Zack-Tay committed:
add mark as done for task in Fredicksen.java Level-3

Zack-Tay committed:
created new Task.java to encapsulate each Task and completed Level-4

Zack-Tay committed:
add testcases for Fredricksen.java and completed A-TextUiTesting

Zack-Tay committed:
add error message for some invalid inputs to Fredricksen.java and completed Level-5

Zack-Tay committed:
add delete task feature to Fredricksen.java and completed Leve;-6

Zack-Tay committed:
add enum for list of commands for Fredricksen.java and Completed A-enums

Zack-Tay committed:
Add loadlist and updateList and completed Level-7

Zack-Tay committed:
Merge branch 'branch-Level-7'

Zack-Tay committed:
Add DateTime feature and Completed Level-8

Zack-Tay committed:
Merge branch 'branch-Level-8'

Zack-Tay committed:
Add OOP abstractions and Completed A-MoreOOP

Zack-Tay committed:
Merge branch 'branch-A-MoreOOP'

Zack-Tay committed:
Add packages and Completed A-Packages

Zack-Tay committed:
Merge branch 'branch-A-Packages'

Zack-Tay committed:
Merge branch 'add-gradle-support' of https://github.com/Zack-Tay/ip

* 'add-gradle-support' of https://github.com/Zack-Tay/ip:
  Bump gradle and lib version
  Add Gradle support
  docs/README.md: Tweak document template

Zack-Tay committed:
Add JUnit UiTest and TaskTest

Zack-Tay committed:
Merge pull request #1 from Zack-Tay/branch-A-JUnit

Add JUnit UiTest and TaskTest

Zack-Tay committed:
Add Junit UiTest and TaskTest

Zack-Tay committed:
Merge branch 'branch-A-JUnit'

Zack-Tay committed:
Add Javadocs to public methods and classes

Zack-Tay committed:
Add Coding standards and checkstyle

Zack-Tay committed:
Add find command for keywords of tasks

Zack-Tay committed:
Merge branch 'branch-A-JavaDoc'

Zack-Tay committed:
Merge branch 'branch-A-CodingStandard'

Zack-Tay committed:
Merge branch 'branch-Level-9'

linnn-7 commented on src/main/java/Fredricksen.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    public void run() {
+        ui.showWelcome();
+        ui.listOfCommands();
+        System.out.println("");
+        boolean isLoop = true;
+        // while loop
+        while(isLoop) {
+            String task = ui.readCommand();
+            String[] splitTask = task.split(" ");
+            if (splitTask[0].equals("bye")) {
+                isLoop = false;
+            } else {
+                Command command = new Command(task);
+                command.execute(task, list, ui, store);
+                store.updateFile(list);
+            }
+        }
+        ui.closeScanner();
```
I like how you design your OOP structure such that your execution path is very clear and prominent!

linnn-7 commented on src/main/java/commands/Command.java:
```diff
@@ -0,0 +1,273 @@
+    }
+
+    public String getTaskSymbol(String command) {
+        String res = "";
+        switch (command) {
+        case "todo":
+            res = "T";
+            break;
+        case "event":
+            res = "E";
+            break;
+        case "deadline":
+            res = "D";
+            break;
+        default:
+            break;
+        }
```
Maybe can consider using enums here?

Zack-Tay replied:
Okay I will look into that thanks!

linnn-7 commented on src/main/java/commands/Command.java:
```diff
@@ -0,0 +1,273 @@
+            case "todo":
+            case "event":
+            case "deadline":
+                ui.showLine();
+                System.out.println("Got it. I've added this task: ");
+                String symbol = this.getTaskSymbol(command[0]);
+                int byStartIndex = fullCommand.indexOf("/by");
+                int fromStartIndex = fullCommand.indexOf("/from");
+                if (byStartIndex == -1 && fromStartIndex == -1) {
+                    String des = fullCommand.substring(command[0].length() + 1);
+                    Task newTask = new Task(des, symbol, false);
+                    tasks.addTask(newTask);
+                } else if (fromStartIndex == -1) {
+                    byCommand(fullCommand, byStartIndex, symbol, tasks);
+                } else {
+                    fromCommand(fullCommand, fromStartIndex, symbol, tasks);
+                }
+
+                Task latestTask = tasks.getTask(tasks.size() - 1);
+                String task = latestTask.toString();
+                System.out.println("    " + task);
+                String singular = tasks.size() == 1 ? "task" : "tasks";
+                int num = tasks.size();
+                System.out.println("Now you have " + num + " " + singular + " in the list.");
+                ui.showLine();
+                break;
```
Maybe can add helper functions to handle this part so that your execute function can be shorter?

Zack-Tay replied:
Thanks for pointing it out, I have abstracted some of the logic away into a helper function.

linnn-7 commented on src/main/java/ui/Ui.java:
```diff
@@ -0,0 +1,74 @@
+     * and how to use them to get valid outputs.
+     */
+    public void listOfCommands() {
+        System.out.println("Below are the available commands and formats to follow!");
+        System.out.println("1. To view all your current task: list");
+        System.out.println("2. To add new task:");
+        System.out.println("    a. todos: todo <task>");
+        System.out.println("    b. deadlines: deadline <task> /by <deadline>");
+        System.out.println("    c. event: Event <event> /from <startdate, starttiming> /to <enddate, endtiming>");
+        System.out.println("3. To delete a task: delete <task number>");
+        System.out.println("4. To mark task as completed: mark <task number>");
+        System.out.println("5. To unmark completed task: unmark <task number>");
+        System.out.println("6. To exit program: bye");
```
Maybe can use "\n" to indicate a new line

Zack-Tay replied:
I have reduced the number of print statements already, thanks for pointing it out!

linnn-7 reviewed (commented):
I appreciate the thoughtfulness in your project's OOP structure design. While reviewing, I noticed a few minor changes that you might want to consider. Overall, I believe your code is already well-structured and poised for future extensions.

Zack-Tay committed:
Add TaskType package

Zack-Tay committed:
Add TaskType package

Zack-Tay committed:
Merge branch 'branch-refactor'

Zack-Tay committed:
Add checkstyle for class methods

Zack-Tay committed:
Merge branch 'branch-CheckStyle'

Zack-Tay committed:
Add GUI for chatbot

Zack-Tay committed:
Merge branch 'branch-Level-10'

Zack-Tay committed:
Add GUI features with chatbot response

yyccbb commented on src/main/java/Fredricksen.java:
### Package
Should all classes be put into packages? (i.e. you could have a root package named "fredricksen" )

yyccbb commented on src/main/java/Fredricksen.java:
```diff
@@ -0,0 +1,63 @@
+    public void run() {
+        ui.showWelcome();
+        boolean isLoop = true;
+        // while loop
```
### Redundant comment
Is `// while loop` necessary?

yyccbb commented on src/main/java/commands/Command.java:
```diff
@@ -0,0 +1,206 @@
+                output = ui.output(findCommand(matchList));
+                break;
+            default:
+                output = ui.output("Sorry Old Man Fredricksen don't recognise this input! Type \"help\" if you need a guide on input format!");
```
### General
I like it 💯 💯

yyccbb commented on src/test/java/tasks/TaskTest.java:
```diff
@@ -0,0 +1,39 @@
+    /**
+     * Checks if toString method returns string with correct format.
+     */
+    /** @Test
+     public void taskToString_correctFormat() {
+        String timeCommand = "2/12/2019 1800";
+        DateTimeFormatter format = DateTimeFormatter.ofPattern("d/M/yyyy HHmm");
+        LocalDateTime time = LocalDateTime.parse(timeCommand, format);
+        Task newTask = new Task("deadline homework /by 2/12/2019 1800", "D", false, time);
+        assertEquals(newTask.toString(), "[D][] deadline homework /by 2/12/2019 1800");
+    } */
```
### JavaDoc
I know this is temporary but do remember to remove it later :)

yyccbb commented on src/main/java/tasks/taskType/Task.java:
```diff
@@ -0,0 +1,84 @@
+        };
+    }
+
+    public String formatDates(String date) {
+        String result = "";
+        LocalDateTime time = null;
+        LocalDate dateWoTime = null;
+        DateTimeFormatter[] formats = this.getFormatsDateTime();
+        for (DateTimeFormatter format : formats) {
+            try {
+                time = LocalDateTime.parse(date, format);
+                break;
+            } catch (DateTimeParseException err) {
+                dateWoTime = LocalDate.parse(date, format);
+                break;
+            } catch (Exception err) {
+                // System.out.println("too bad");
+            }
+        }
```
### JavaDoc
Consider having JavaDoc for more complicated methods like this?

yyccbb commented on src/main/java/commands/Command.java:
### General
Hey, I like how you implemented your command system. I know this is not part of my job as the first reviewer but I just wanted to give you some ideas if you are interested: consider having a dispatcher class that takes in `String command` and sends out executables that perform an action when their .execute() method is called. You can refer to my codebase if interested :) All these are extra but may help make your codebase look cleaner.

Zack-Tay replied:
Thanks for your comment. Really grateful that you have reviewed my work in good depth and provided very useful suggestions! I will continue to work the recommendations u have provided. Many thanks.

yyccbb commented on src/main/java/ui/Ui.java:
### General
Hey, I like how you implemented your ui system. A small tip might be that some classes can be made singletons and their method can be invoked with `ClassName.method()`. Again, this is extra but might be helpful:)

yyccbb reviewed (commented):
Hi! I like the overall completeness of your project. There are some minor issues with JavaDoc but everything else looks good to me. Good job 💯

Zack-Tay committed:
Add GUI interaction with chatbot

Zack-Tay committed:
Add Command executables

Zack-Tay committed:
Add more commands

Zack-Tay committed:
Add more OOP

Zack-Tay committed:
Merge tag 'A-CodeQuality'

Zack-Tay committed:
Add Assertion for assumptions

Zack-Tay committed:
Merge branch 'branch-A-Assertions'

Zack-Tay committed:
Add streams

Zack-Tay committed:
Merge branch 'branch-A-Streams'

Zack-Tay committed:
Add git commit message body

Add git commit message body practice.

This is to familiarise myself with the process and structure of
adding git commit message body.

It is done from SourceTree as it is convenient and easy.

Let's
* Add an assert statement to ensure that task is not empty when
passed into the execute function.

Zack-Tay committed:
Merge branch 'branch-A-FullCommitMessage'

Zack-Tay committed:
Add Cat C-MassOps feature

Add a new feature where users can perform some mass operations on
all their current tasks.

This makes it easier for the user to do a mass operation and the
user do not have to manually say mark each task as done at
the end of the day, but rather can just do 1 mass operation to
mark all the tasks as done.

Let's,
* Update DeleteCommand.java to support mass delete of all tasks
currently in the TaskList.
* Update MarkCommand.java to support mass mark of all tasks
currently in the TaskList.
* Update UnmarkCommand.java to support mass unmark of all tasks
currently in the TaskList.

Zack-Tay committed:
Merge branch 'branch-A-BCDExtension'

Zack-Tay committed:
Add better GUI

Zack-Tay committed:
Add more styling and better GUI

Zack-Tay committed:
Merge branch 'branch-A-BetterGui'

Zack-Tay committed:
Add more testing and refactor code

Zack-Tay committed:
Merge branch 'branch-A-MoreTesting'

Zack-Tay committed:
Add javadocs and coding standards

Zack-Tay committed:
Merge branch 'branch-refactorCode'

Zack-Tay committed:
Add user guide and Ui png

Zack-Tay committed:
Merge branch 'branch-A-UserGuide'

Zack-Tay committed:
Add mass ops feature cat C

Zack-Tay committed:
Merge branch 'branch-BCD-Extension'
